
diff --git a/lib/utils.c b/lib/utils.c
index 4e9c719..eb92414 100644
--- a/lib/utils.c
+++ b/lib/utils.c
@@ -487,7 +487,7 @@ int get_addr(inet_prefix *dst, const char *arg, int family)
 		exit(1);
 	}
 	if (get_addr_1(dst, arg, family)) {
-		fprintf(stderr, "Error: an inet address is expected rather than \"%s\".\n", arg);
+		fprintf(stderr, "Error: an %s address is expected rather than \"%s\".\n", (family == AF_INET6 ? "inet6" : "inet") ,arg);
 		exit(1);
 	}
 	return 0;
@@ -500,7 +500,7 @@ int get_prefix(inet_prefix *dst, char *arg, int family)
 		exit(1);
 	}
 	if (get_prefix_1(dst, arg, family)) {
-		fprintf(stderr, "Error: an inet prefix is expected rather than \"%s\".\n", arg);
+		fprintf(stderr, "Error: an %s prefix is expected rather than \"%s\".\n", (family == AF_INET6 ? "inet6" : "inet") ,arg);
 		exit(1);
 	}
 	return 0;
