--- iproute2-2.6.31/ip/link_veth.c	2009-12-26 19:26:44.000000000 +0100
+++ iproute2-2.6.31/ip/link_veth.c.new	2011-02-04 13:28:01.020485877 +0100
@@ -27,7 +27,10 @@
 static int veth_parse_opt(struct link_util *lu, int argc, char **argv,
 			  struct nlmsghdr *hdr)
 {
-	char *name, *type, *link, *dev;
+	char *name = NULL;
+	char *type = NULL;
+	char *link = NULL;
+	char *dev = NULL;
 	int err, len;
 	struct rtattr * data;
 
