diff --git a/lib/libnetlink.c b/lib/libnetlink.c
index 6771d99..8f6bb75 100644
--- a/lib/libnetlink.c
+++ b/lib/libnetlink.c
@@ -99,19 +99,19 @@ int rtnl_wilddump_req_filter(struct rtnl_handle *rth, int family, int type,
 {
 	struct {
 		struct nlmsghdr nlh;
-		struct ifinfomsg ifm;
-		/* attribute has to be NLMSG aligned */
-		struct rtattr ext_req __attribute__ ((aligned(NLMSG_ALIGNTO)));
+		struct rtgenmsg g;
+		__u16 align_rta;        /* attribute has to be 32bit aligned */
+		struct rtattr ext_req;
 		__u32 ext_filter_mask;
 	} req;
 
 	memset(&req, 0, sizeof(req));
 	req.nlh.nlmsg_len = sizeof(req);
 	req.nlh.nlmsg_type = type;
-	req.nlh.nlmsg_flags = NLM_F_DUMP|NLM_F_REQUEST;
+	req.nlh.nlmsg_flags = NLM_F_ROOT|NLM_F_MATCH|NLM_F_REQUEST;
 	req.nlh.nlmsg_pid = 0;
 	req.nlh.nlmsg_seq = rth->dump = ++rth->seq;
-	req.ifm.ifi_family = family;
+	req.g.rtgen_family = family;
 
 	req.ext_req.rta_type = IFLA_EXT_MASK;
 	req.ext_req.rta_len = RTA_LENGTH(sizeof(__u32));
