diff --git a/include/linux/if_link.h b/include/linux/if_link.h
index 52a5044..6f918df 100644
--- a/include/linux/if_link.h
+++ b/include/linux/if_link.h
@@ -83,7 +83,14 @@ enum {
 	IFLA_STATS64,
 	IFLA_VF_PORTS,
 	IFLA_PORT_SELF,
+	IFLA_AF_SPEC,
+	IFLA_GROUP,		/* Group the device belongs to */
+	IFLA_NET_NS_FD,
 	IFLA_EXT_MASK,          /* Extended info mask, VFs, etc */
+	IFLA_PROMISCUITY,       /* Promiscuity count: > 0 means acts PROMISC */
+#define IFLA_PROMISCUITY IFLA_PROMISCUITY
+	IFLA_NUM_TX_QUEUES,
+	IFLA_NUM_RX_QUEUES,
 	__IFLA_MAX
 };
 
